### DISCUSSION TITLE
Cannot set 'isTouched' to a field inside Controller

### DISCUSSION THREAD

**Discussion by sergeyivanovv89** [react-hook-form/react-hook-form#6824] [Q&A]
### Version Number

7.17.4

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-possibility-set-istouched-inside-controller-0wg08

### Steps to reproduce

1. Set date via date picker on the first input
2. 'touchedFields' object is empty now

### Expected behaviour

When we deal with components of third-party libraries such as MUI DatePicker we need to set 'isTouched' to a field when interacting with it. Now we can't.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
![Screen Shot 2021-10-15 at 8 40 24 pm](https://user-images.githubusercontent.com/10513364/137467179-0114d2fa-59c4-4819-9d6d-61c6e880dda7.png)

This is not an issue with hook form, looks like the component doesn't have `onBlur` prop, you probably know better than i when to fire onBlur with this component, as long as `onBlur` is executed, hook form will register the input as touched.

**Comment by sergeyivanovv89** [comment]
Oh, in previous versions of react-hook-form it all worked as you described. I've updated my codesandbox and now you can see strange behavior. 
1. Set a date via date picker. The date is not empty but the field error appeared. 
2. Set a date a second time. There's no error now.

If you downgrade to version for example 7.11.0 it all works correctly.

**Comment by bluebill1049** [comment]
then the old version must had a bug then, because it shouldn't update touchedFields unless onBlur is invoked.

**Comment by sergeyivanovv89** [comment]
The old version doesn't have a bug:
https://codesandbox.io/s/react-hook-form-possibility-set-istouched-inside-controller-forked-kzqy5
But the new one does, see the [comment](https://github.com/react-hook-form/react-hook-form/issues/6816#issuecomment-944167867) above.

**Comment by sergeyivanovv89** [comment]
In the new version we can see validation bug when calling onBlur()

**Comment by bluebill1049** [comment]
the old one had a bug that takes in `onChange` and updated the `touchedFields`, the latest version corrected that.` onBlur` will need to be invoked to update the `touchedFields`. 

> In the new version we can see validation bug when calling onBlur()

what's the validation bug?

**Comment by sergeyivanovv89** [comment]
https://www.loom.com/share/00cbfe44a6c94dfab2c4150e6cf66838

**Comment by sergeyivanovv89** [comment]
![image](https://user-images.githubusercontent.com/22203394/137484118-90e3b27e-075d-43ed-8bff-d97a64e205e8.png)

Invoking `onBlur()` with argument gives the same result as with `undefined`

**Comment by bluebill1049** [comment]
fixed and please refer to my example: https://codesandbox.io/s/react-hook-form-possibility-set-istouched-inside-controller-forked-042m5?file=/src/App.jsx and not invoking onBlur and onChange at the same function.

```diff
<Controller
  name="date"
  rules={{ required: "Required" }}
  render={({
    field: { ref, ...fieldProps },
    fieldState: { error }
  }) => (
    <>
      <DatePicker
        {...fieldProps}
-       onChange={(date) => {
-          fieldProps.onChange(date);
-          fieldProps.onBlur();
-        }}
-        renderInput={(props) => <TextField {...props} />}
+       renderInput={(props) => (
+         <TextField {...props} onBlur={fieldProps.onBlur} />
+       )}
      />
      {error && <p>{error.message}</p>}
    </>
  )}
/>
```

**Comment by sergeyivanovv89** [comment]
In your example 'isTouched' is not set after selecting date via picker

**Comment by bluebill1049** [comment]
now it does: https://codesandbox.io/s/react-hook-form-possibility-set-istouched-inside-controller-forked-042m5?file=/src/App.jsx

**Comment by sergeyivanovv89** [comment]
Thanks a lot. You're cool

**Comment by sergeyivanovv89** [comment]
https://codesandbox.io/s/react-hook-form-possibility-set-istouched-inside-controller-forked-serod
Please try to fill date via picker and in console you can see that there's no validation error initially and then the error appears and finally there's no error again. This situation causes the interface to blink (((

**Reply by bluebill1049** [reply]
can you share a video?

**Reply by sergeyivanovv89** [reply]
Here you are
https://www.loom.com/share/abe53bee35c540d783e606e69903ff31

**Reply by sergeyivanovv89** [reply]
It seems to be a bug.

**Reply by bluebill1049** [reply]
it's not a bug with hook form, instead of the complication between the click calendar icon, the original onBlur is when an input gets blued, I think if you want to treat this input as touched, perhaps onBlur need to fire when the user selects on the calender icon too. How hook form determines an input is touched when `blur` even happened.

**Comment by bluebill1049** [comment]
perhaps try this approach instead: 

```tsx
<Controller
  name="date"
  rules={{ required: "Required" }}
  render={({
    field: { ref, ...fieldProps },
    fieldState: { error }
  }) => {
    console.log(error);
    return (
      <>
        <DatePicker
          {...fieldProps}
          onChange={(date) => {
            fieldProps.onChange(date);
          }}
          onClose={() => fieldProps.onBlur()}
          renderInput={(props) => (
            <TextField
              {...props}
              onBlur={() => {
                fieldProps.onBlur();
              }}
            />
          )}
        />
        {error && <p>{error.message}</p>}
      </>
    );
  }}
/>
```

**Reply by scottwernervt** [reply]
Solution above resolved my issue with validation not re-running after selecting a date. Note using `react-hook-form@v6`.

```jsx
<Controller
  name="releaseDate"
  control={control}
  rules={{
    required: {
      value: true,
      message: "This is a required field.",
    },
  }}
  render={(props) => {
    return (
      <ReactDatePicker
        id={props.name}
        name={props.name}
        onChange={props.onChange}
        onBlur={props.onBlur}
        onCalendarClose={() => props.onBlur()} // trigger validation
        selected={props.value}
        isClearable
      />
    );
  }}
/>;

```

https://codesandbox.io/s/react-hook-form-v6-reactdatepicker-76r72?file=/src/index.js
